### ISSUE TITLE
Intermittent segfaults during libtorch shutdown + Proposed Fix

### ISSUE THREAD

**Issue by VivekPanyam** [pytorch/pytorch#36408]
## 🐛 Bug

We have several binaries that use libtorch and a small number of them have started to segfault during shutdown. These segfaults are intermittent and happen at different rates when running different models.

A detailed description of the issue, the cause, and a proposed fix are included below.

## Description of Issue

Most of the backtraces we got weren't very helpful, but a few included some useful information:
```
0x00007fffa701ac58 in std::default_delete<at::cuda::(anonymous namespace)::DeviceThreadHandlePool<cudnnContext*, &at::native::(anonymous namespace)::createCuDNNHandle, &at::native::(anonymous namespace)::destroyCuDNNHandle>::PoolWindow>::operator()(at::cuda::(anonymous namespace)::DeviceThreadHandlePool<cudnnContext*, &at::native::(anonymous namespace)::createCuDNNHandle, &at::native::(anonymous namespace)::destroyCuDNNHandle>::PoolWindow*) const ()
```

```
0x00007fffa700ec38 in std::default_delete<at::cuda::(anonymous namespace)::DeviceThreadHandlePool<cublasContext*, &at::cuda::(anonymous namespace)::createCublasHandle, &at::cuda::(anonymous namespace)::destroyCublasHandle>::PoolWindow>::operator()(at::cuda::(anonymous namespace)::DeviceThreadHandlePool<cublasContext*, &at::cuda::(anonymous namespace)::createCublasHandle, &at::cuda::(anonymous namespace)::destroyCublasHandle>::PoolWindow*) const ()
```

These appear to correspond to the destruction of `myPoolWindow` in the below snippets:

https://github.com/pytorch/pytorch/blob/8cb19508059332329a7993b34d92b46080673a71/aten/src/ATen/cudnn/Handle.cpp#L25-L47

https://github.com/pytorch/pytorch/blob/8cb19508059332329a7993b34d92b46080673a71/aten/src/ATen/cuda/CublasHandlePool.cpp#L24-L45

## Cause of the segfault

The cause of the segfault is a combination of a few things:

1. `PoolWindow` is initialized with a reference to a `DeviceThreadHandlePool` (`pool` in the snippets above) and it attempts to use this reference during destruction:

https://github.com/pytorch/pytorch/blob/8cb19508059332329a7993b34d92b46080673a71/aten/src/ATen/cuda/detail/DeviceThreadHandles.h#L83-L88

https://github.com/pytorch/pytorch/blob/8cb19508059332329a7993b34d92b46080673a71/aten/src/ATen/cuda/detail/DeviceThreadHandles.h#L120-L127

2. `PoolWindow` can outlive the `DeviceThreadHandlePool` it was initialized with.

In the snippets at the top of this issue, `myPoolWindow` is defined as `thread_local`. This means one instance will be created per thread that calls `getCudnnHandle()` or `getCurrentCUDABlasHandle()`

The problem is that `thread_local` objects are destroyed when the calling thread exits.

Since it's possible for a thread to be defined in another translation unit and [initialization/destruction order across translation units is unspecified](https://isocpp.org/wiki/faq/ctors#static-init-order), it is possible for calling threads to exit after `pool` is destroyed.

This means that `myPoolWindow` could be destroyed after `pool`, causing `release()` to use an invalid reference.

## Proposed Fix

Refactor `DeviceThreadHandlePool` to inherit from `std::enable_shared_from_this` and pass a shared pointer into to the `PoolWindow` constructor instead of a reference.

This fixed the issue in our internal build of Torch. Happy to put up a PR if you're okay with this fix.

_Note: this issue also exists in `aten/src/ATen/cuda/CuSparseHandlePool.cpp` although we didn't see any segfaults there in our applications._

cc @ezyang @gchanan @zou3519 @ngimel

**Comment by ngimel** [comment]
cc @mcarilli

**Comment by ngimel** [comment]
Setting high priority because of a segfault.

**Comment by ngimel** [comment]
Please put up a PR, it's easier to have a discussion there.

**Comment by mcarilli** [comment]
I think this is a compiler bug (https://github.com/pytorch/pytorch/pull/36416#issuecomment-612294692).

**Comment by VivekPanyam** [comment]
Thanks for the taking a look! I posted a comment with details and counterexamples showing that this likely isn't a compiler bug (https://github.com/pytorch/pytorch/pull/36416#issuecomment-612575419)

Thanks again for getting to this so quickly!

**Comment by ezyang** [comment]
cc @malfet

**Comment by peterbell10** [comment]
Looks like this was fixed in #36416
